wire-kotlin-generator/src/main/java/com/squareup/wire/kotlin/KotlinGenerator.kt

```diff
@@ -0,0 +1,399 @@
+    }
+
+    return FunSpec.builder("encodedSize")
+        .addParameter("value", className)
+        .returns(Int::class.asTypeName())
+        .addCode(body
+            .add("value.unknownFields.size()\n")
+            .build())
+        .addModifiers(OVERRIDE)
+        .build()
+  }
+
+  private fun encodeFunc(message: MessageType): FunSpec {
+    val className = generatedTypeName(message)
+    var body = CodeBlock.builder()
+
+    message.fields().forEach { field ->
+      val adapterName = getAdapterName(field)
+      body.addStatement("%L.%LencodeWithTag(writer, %L, value.%L)",
+          adapterName, if (field.isRepeated) "asRepeated()." else "", field.tag(), field.name())
+    }
+
+    body.addStatement("writer.writeBytes(value.unknownFields)")
+
+    return FunSpec.builder("encode")
+        .addParameter("writer", ProtoWriter::class.asClassName())
+        .addParameter("value", className)
+        .addCode(body.build())
+        .addModifiers(OVERRIDE)
+        .build()
+  }
+
+  private fun decodeFunc(message: MessageType): FunSpec {
+    val className = nameToKotlinName[message.type()]!!
+
+    var declarationBody = CodeBlock.builder()
+
+    var returnBody = CodeBlock.builder()
+    returnBody.add("return %L(\n", className.simpleName)
+
+    var decodeBlock = CodeBlock.builder()
+    decodeBlock.addStatement("val unknownFields = reader.decodeMessage { tag -> ")
+
+    val INDENTATION = " ".repeat(4)
+
+    // Indent manually as code generator doesn't handle this block gracefully.
+    decodeBlock.addStatement("%Lwhen(tag) {", INDENTATION)
+    val nameAllocator = nameAllocator(message)
+
+    val missingRequiredFields = ClassName("com.squareup.wire.internal", "Internal")
+
+    message.fields().forEach { field ->
+      val adapterName = getAdapterName(field)
+      var throwExceptionBlock = ""
+      var declarationBodyTemplate = ""
+      var decodeBodyTemplate = ""
+      val fieldName = nameAllocator.get(field)
+      var fieldClass = nameToKotlinName[field.type()]!!
+
+      if (field.isRepeated) {
+        declarationBodyTemplate = "var %L = mutableListOf<%L>()"
+        decodeBodyTemplate = "%L%L -> %L.add(%L.decode(reader))"
+      } else {
+        declarationBodyTemplate = "var %L: %L = null"
+        fieldClass = fieldClass.asNullable()
+        decodeBodyTemplate = "%L%L -> %L = %L.decode(reader)"
+
+        if (!field.isOptional) {
+          throwExceptionBlock = CodeBlock.of(" ?: throw %T.%L(%L, \"%L\")",
+              missingRequiredFields, "missingRequiredFields", field.name(), field.name()).toString()
+        }
+      }
+
+      declarationBody.addStatement(declarationBodyTemplate, fieldName, fieldClass)
+      decodeBlock.addStatement(decodeBodyTemplate, INDENTATION.repeat(2), field.tag(), fieldName,
+          adapterName)
+      returnBody.add("%L%L = %L%L,\n", INDENTATION, fieldName, fieldName, throwExceptionBlock)
```

Comment:
...so that when inserted here you still get imports. I believe you can still use `%L` with `CodeBlock`s